stdlib/option.ml

```diff
@@ -0,0 +1,39 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*   Copyright 2018 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+type 'a t = 'a option
+
+let some v = Some v
+let evict ~none = function Some v -> v | None -> none
+let get = function Some v -> v | None -> invalid_arg "option is None"
+let map f o = match o with None -> None | Some v -> Some (f v)
+let flat_map f o = match o with None -> None | Some v -> f v
+let flatten = function Some (Some _ as o) -> o | _ -> None
```

Comment:
`flat_map` is also known as monadic bind.  Could we make sure this Option module will work with whatever PPX implementing monads we have today?  And perhaps with a future, long-overdue "monadic let" syntax?

Reply:
I can certainly add a `bind` and a `return` function if that's desired, but then which signature should be supported ? 

A quick search in `opam` suggest there are already at least four `ppx`s to this: `ocaml-monadic`, `omonad`,  `ppx_monadic` and `ppx_let` and as far as I can tell the interface needed by the latter is incompatible with the interface of the others.

Couldn't we simply add these name aliases in a `Monad` submodule once this future, long-overdue "monadic let" syntax lands ?

Reply:
Can I suggest just removing `flat_map` from this PR? The name really comes from the equivalent function on lists, but we don't actually have that function in the standard library. By removing it we can avoid getting into issues about how people want the standard library to deal with monads and other computation abstractions on a PR that is about bringing support for options into line with support for other built-in types.

Reply:
@lpw25 I don't understand your concern and request:

1.  `flat_map` as defined in this PR would in any case not be a good candidate for a monadic interface due to the order in which I wrote it (`bind` has the argument swapped). 
2. I wouldn't mind if `List.flat_map` was actually added to the stdlib. 
3. I don't see why adding these functions poses a problem and how not adding these useful functions "avoid[s] getting into issues about how people want the standard library to deal with monads and other computation abstractions ".

I mean at that rate why don't you tell me to remove `Option.flatten`, this is just monadic `join` in disguise.

Reply:
> I wouldn't mind if List.flat_map was actually added to the stdlib.

I have no particular opinion on whether `List.flat_map` should be added or not -- my point is just that I would rather see `List.flat_map` and `Option.flat_map` added together in a different PR -- where discussions about how they should named and the order of their arguments can be discussed. Essentially, I find it very strange to have `Option.flat_map` without `List.flat_map` because the name `flat_map` derives from the list version of the function.

> I mean at that rate why don't you tell me to remove Option.flatten

Because we already have `List.flatten` so adding `Option.flatten` is consistent.

Reply:
Not really convinced but I'm a pleasing machine this evening. Function removed from the PR.

Reply:
(and for interested persons I have no plan to carry the work for `{List,Option}.flat_map` myself)